### PR TITLE
UnnecessaryParentheses: add options to allow in ambiguous cases

### PR DESCRIPTION
**Pull request by dzirbel**
One blocker for adopting the UnnecessaryParentheses is that in some common cases parentheses that may not be strictly required by order of operations are still enormously helpful in parsing code, such as for boolean operations like `(x && y) || z`. While it's possible to `@Suppress` each of these, this adds bloat and encourages developers to avoid these (in my opinion) often very helpful indicators of code flow.

Instead, we can add a configuration option to UnnecessaryParentheses to skip over some of the most common cases where unnecessary parentheses can be useful in clarifying the code. This strategy was adopted from IntelliJ's UnclearPrecedenceOfBinaryExpressionInspection inspection: https://github.com/JetBrains/intellij-community/blob/master/plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/UnclearPrecedenceOfBinaryExpressionInspection.kt with a fair amount of trimming and a different set of operators that have unclear precedence (adding numeric and boolean ones).

This idea has been discussed a number of times; most recently in #4495, which suggested that it could integrated into detekt (although perhaps only by porting the functionality from IntelliJ). Previous issues such as #2789 and #1929 concluded that it would be too much of an investment or overcomplicate the rule. I'm happy to stick with that precedent, but as a user who doesn't want to adopt this rule on my team until we can avoid at least some of the more common cases where parentheses can be helpful I thought I'd suggest a concrete implementation with fairly low complexity (although admittedly only addressing some of the lowest hanging fruit). I haven't thought through every edge case or application, so perhaps complexity would explode with a more fleshed-out version.

### CODE REVIEW

File: detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryParenthesesSpec.kt [hunk 1]
```diff
@@ -184,4 +188,46 @@ class UnnecessaryParenthesesSpec {
...
+
+        @ParameterizedTest
+        @ValueSource(booleans = [true, false])
+        fun `does not report numeric expressions`(allow: Boolean) {
+            val config = TestConfig(mapOf("allowInAmbiguousNumericExpressions" to allow))
+            val code = "val local = (1 * 2) + 3"
+
+            assertThat(UnnecessaryParentheses(config).lint(code)).hasSize(if (allow) 0 else 1)
+        }
+    }
```
Comment (BraisGabin) [inline detekt-rules-style/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryParenthesesSpec.kt:232]: What would happen with this cases: 

- `val local = (1 * 2)`
- `val local = (1 * 2 + 3)`
- `val local = (a || b)`
- `val local = (a || b && c)`
- `val local = (a || b) || c`

To me all of them should be flagged always.

File: detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryParentheses.kt [hunk 2]
```diff
@@ -46,17 +51,48 @@ class UnnecessaryParentheses(config: Config = Config.empty) : Rule(config) {
         Debt.FIVE_MINS
     )
 
+    @Configuration("whether to allow for potentially ambiguous boolean operations such as (x && y) || z")
+    private val allowInAmbiguousBooleanExpressions: Boolean by config(defaultValue = false)
+
+    @Configuration("whether to allow for potentially ambiguous numeric operations such as (x * y) + z")
+    private val allowInAmbiguousNumericExpressions: Boolean by config(defaultValue = false)
```
Comment (cortinico) [inline detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/UnnecessaryParentheses.kt:58]: I think naming is (sadly) ambiguous.
It's not immediately clear what's the semantic of `AmbiguousBooleanExpressions` and `AmbiguousNumericExpressions` and we're forcing the user to read the example.

I would try to find a better name if possible.

**Comment by BraisGabin** [comment]
> Sorry for the noise in the meantime.

It's not noise! Thanks for proposing this and create a first draft to give fast feedback.

This is really help us. It doesn't feel good to decline or force someone to change nearly all the code that s/he work for weeks. But we need to do those type of things to keep the project healthy and "managable". So don't worry about sending drafts of ideas we always welcome any help.

**Comment by cortinico** [comment]
> Sorry for the noise in the meantime.

Don't worry about that :) We're more than happy to receive contribs like yours.

**Comment by dzirbel** [comment]
OK, I have a new approach heavily modified from IntelliJ's inspection https://github.com/JetBrains/intellij-community/blob/master/plugins/kotlin/idea/src/org/jetbrains/kotlin/idea/inspections/UnclearPrecedenceOfBinaryExpressionInspection.kt which is more generic and should (I believe) cleanly cover cases where parentheses are not strictly required but can be very helpful for clarifying the meaning of the code.

This might even open the door for the inverse warning matching the behavior of IntelliJ's inspection: warning when parentheses could be helpful to clarify precedence but aren't currently present. I don't feel strongly about that idea - probably wouldn't use it myself but perhaps others would.

Is there a standard way of unit testing different rule configurations that I should use here? I wanted to test all the existing cases with the new option applied to make sure it wouldn't change their behavior, but this added a fair bit of boilerplate. Perhaps `@ParameterizedTest` could clean those up but would make it hard for the tests that have different behavior between configurations.

**Comment by BraisGabin** [comment]
I think that to test this properly you could move that code out of the rule. And test the function itself. There you can use a parameterized test to add all the cases you want. Given x code the function should return (true/false). 

Then you integrate that function inside the rule and you don't need to test all the cases there, just some to check that the integration is correct.

**Comment by dzirbel** [comment]
Good ideas - I've refactored a bit and went with a different testing strategy using `@ParameterizedTest` and providing a different instance of the rule while also making it easy to check whether `allowForUnclearPrecedence` is active for this test case. This makes most tests easy and avoids needing the extract the functionality or test the integration indirectly. Happy to change that up if you feel that another way to be preferred though.

**Review by BraisGabin — APPROVED** [review]
I like it :)

There are a lot of tests but I think that's good because there are a lot of cases and I think that you are cheking all of them.

Really good job 👍

**Comment by BraisGabin** [comment]
I'm moving this to `1.22.0` because it feels like a big change and we are near to release a stable version but as soon as we release `1.21.0` this will be merged.

**Comment by TWiStErRob** [comment]
This got me, the documentation says the flag exists, but when trying to use it, it fails to execute. Is this normal @BraisGabin that the docs don't match the latest release?

**Comment by TWiStErRob** [comment]
@dzirbel I would like to throw another case at you where clarifying parentheses might help:
```kotlin
private fun fakeClock(fixture: JFixture, minOffset: Long, maxOffset: Long): Clock? {
	val now = System.currentTimeMillis()
	val millis = fixture.buildRange((now - minOffset)..(now + maxOffset))
	val zone = fixture.create().fromList(*ZoneId.getAvailableZoneIds().toTypedArray())
	return Clock.fixed(Instant.ofEpochMilli(millis), ZoneId.of(zone))
}
```
Without parens `fixture.buildRange(now - minOffset..now + maxOffset)` one might expect it's interpreted as `fixture.buildRange(now - (minOffset..now) + maxOffset)`

**Comment by dzirbel** [comment]
> This got me, the documentation says the flag exists, but when trying to use it, it fails to execute. Is this normal @BraisGabin that the docs don't match the latest release?

I believe that is the current behavior, yes - new rules, params, etc appear on the website immediately after they are merged rather than when a new version is released. I've always thought a version dropdown like [Gradle](https://docs.gradle.org/current/userguide/userguide.html) and others use could be helpful.

> @dzirbel I would like to throw another case at you where clarifying parentheses might help:
> 
> ```kotlin
> private fun fakeClock(fixture: JFixture, minOffset: Long, maxOffset: Long): Clock? {
> 	val now = System.currentTimeMillis()
> 	val millis = fixture.buildRange((now - minOffset)..(now + maxOffset))
> 	val zone = fixture.create().fromList(*ZoneId.getAvailableZoneIds().toTypedArray())
> 	return Clock.fixed(Instant.ofEpochMilli(millis), ZoneId.of(zone))
> }
> ```
> 
> Without parens `fixture.buildRange(now - minOffset..now + maxOffset)` one might expect it's interpreted as `fixture.buildRange(now - (minOffset..now) + maxOffset)`

I agree, that's a good case. Opened #5143 to add it.

**Comment by BraisGabin** [comment]
> This got me, the documentation says the flag exists, but when trying to use it, it fails to execute. Is this normal @BraisGabin that the docs don't match the latest release?

Yes, we need a versioned documentation. We know that docusaurus let us handle it but we didn't have the time for it: #4168. Any help there is more than appreciatted.

**Comment by cortinico** [comment]
> > This got me, the documentation says the flag exists, but when trying to use it, it fails to execute. Is this normal @BraisGabin that the docs don't match the latest release?
> 
> Yes, we need a versioned documentation. We know that docusaurus let us handle it but we didn't have the time for it: #4168. Any help there is more than appreciatted.

I can take a stab at #4168
